### PR TITLE
table: fix a bug in TryGetHandleRestoredDataWrapper which may cause panic

### LINKED ISSUE #23805: insert ignore panic on empty clustered index table with new collation enabled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
 CREATE TABLE `tbl_5` (
  `col_25` time NOT NULL DEFAULT '05:35:58',
  `col_26` blob NOT NULL,
  `col_27` double NOT NULL,
  `col_28` char(83) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `col_29` timestamp NOT NULL,
  `col_30` varchar(36) COLLATE utf8mb4_general_ci NOT NULL DEFAULT 'ywzIn',
  `col_31` binary(85) DEFAULT 'OIstcXsGmAyc',
  `col_32` datetime NOT NULL DEFAULT '2024-08-02 00:00:00',
  PRIMARY KEY (`col_26`(3),`col_27`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `idx_10` (`col_26`(5))
);
insert ignore into tbl_5 set col_28 = 'ZmZIdSnq' , col_25 = '18:50:52.00' on duplicate key update col_26 = 'y';
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

ERROR 1105 (HY000): runtime error: index out of range [0] with length 0

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #23805 

Problem Summary:
Didn't skip the column which doesn't need restored data.

### What is changed and how it works?

Skip the column which doesn't need restored data.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

-

### Release note 

- fix a bug in TryGetHandleRestoredDataWrapper which may cause panic

### CODE REVIEW

**Review by tangenta — CHANGES_REQUESTED** [review]
The length of `rsData` is not necessarily equal to those of `pkIdx.Columns`
```
tablecodec.TruncateIndexValue(&rsData[i], useIdx, t.Meta().Columns[idxCol.Offset])
```

**Review by tangenta — APPROVED** [review]

File: table/tables/tables.go [hunk 1]
```diff
@@ -1700,56 +1700,40 @@ func TryGetHandleRestoredDataWrapper(t table.Table, row []types.Datum, rowMap ma
...
 		for _, idxCol := range idx.Columns {
 			if idxCol.Offset == pkCol.Offset {
-				if idxCol.Length == types.UnspecifiedLength || pkCol.Length == types.UnspecifiedLength {
-					break
-				}
-				useIdx := idxCol
-				if pkCol.Length > idxCol.Length {
-					useIdx = pkCol
+				if idxCol.Length == types.UnspecifiedLength || idxCol.Length > truncateTargetCol.Length {
+					truncateTargetCol = idxCol
```
Comment (coocood) [inline table/tables/tables.go:1725]: If `truncateTargetCol.Length == types.UnspecifiedLength`,  truncateTargetCol should still be pkIdxCol

**Pull request merged**